API/ev3_timer.c

```diff
@@ -12,268 +16,552 @@
-  callBack1s_count    = 0;
-
-  // Ensure all pointers are NULL
-  memset(callBack10ms,  0, sizeof(TimerCallback) * MAX_CALLBACKS);
-  memset(callBack50ms,  0, sizeof(TimerCallback) * MAX_CALLBACKS);
-  memset(callBack100ms, 0, sizeof(TimerCallback) * MAX_CALLBACKS);
-  memset(callBack250ms, 0, sizeof(TimerCallback) * MAX_CALLBACKS);
-  memset(callBack500ms, 0, sizeof(TimerCallback) * MAX_CALLBACKS);
-  memset(callBack1s,    0, sizeof(TimerCallback) * MAX_CALLBACKS);
+    // we can call free even on NULL
+    free(pBlock->elements);
+    // clean everything up
+    pBlock->elements = NULL;
+    pBlock->capacity = 0;
+    pBlock->count = 0;
+    pBlock->enabled = false;
 }
 
-void SetTimerCallback(TimerInterval interval, TimerCallback callback)
+bool tmArrayAdd(CallbackBlock *pBlock, TimerCallback callback)
 {
-  if (callback != NULL)
-  {
-	switch(interval)
-	{
-	  case ti10ms:  callBack10ms[callBack10ms_count++] = callback; break;
-	  case ti50ms:  callBack50ms[callBack50ms_count++] = callback; break;
-	  case ti100ms: callBack100ms[callBack100ms_count++] = callback; break;
-	  case ti250ms: callBack250ms[callBack250ms_count++] = callback; break;
-	  case ti500ms: callBack500ms[callBack500ms_count++] = callback; break;
-	  case ti1sec:  callBack1s[callBack1s_count++] = callback; break;
-	}
-  }
+    // ensure that there is at least one free place
+    if (!tmArrayEnsureReserve(pBlock, 1))
+        return false;
+
+    // and put the element in that place
+    pBlock->elements[pBlock->count++] = callback;
+    return true;
 }
 
-void TimerInit(unsigned long interval_multiplier)
+bool tmArrayRemove(CallbackBlock *pBlock, TimerCallback callback)
 {
-	timer_interval_mutliplier = interval_multiplier;
-
-  unsigned long long csTick, msTick, usTick;
-  int i;
-  usTick = TimerGetUS();
-  msTick = usTick / 1000;
-  csTick = msTick / 10;
-  for (i=0; i < NUM_US_TIMERS; i++)
-	usTimers[i] = usTick;
-  for (i=0; i < NUM_MS_TIMERS; i++)
-	msTimers[i] = msTick;
-  for (i=0; i < NUM_CS_TIMERS; i++)
-	csTimers[i] = csTick;
-  // also initialize our callback timers
-  _timerCallbackInit();
+    // look where the callback is stored
+    for (size_t i = 0; i < pBlock->count; i++)
+    {
+        if (pBlock->elements[i] == callback)
+        {
+            // we found it, move succeeding elements an element back
+            memmove(&pBlock->elements[i],
+                    &pBlock->elements[i + 1],
+                    pBlock->count - i - 1);
+            // reduce count by one
+            pBlock->count -= 1;
+            // and return success
+            return true;
+        }
+    }
+    // no element found, removal failed
+    errno = ENOENT;
+    return false;
 }
 
-#endif
+bool tmArrayMirror(CallbackBlock *pDest, CallbackBlock *pSrc)
+{
+    // let's make sure there's enough room to copy the array
+    if (!tmArrayEnsureSize(pDest, pSrc->count))
+        return false;
+
+    // copy trivial fields (but do NOT copy the capacity field)

```

Comment:
Just copy everything and set capacity explicitly to zero?

Reply:
I don't understand the purpose of that. The intent behind preserving the capacity (and the pointer, I forgot to add that to the comment) is to cache previously allocated memory. Setting capacity to zero would trigger a reallocation the next time this function is called.

However I agree that manual copying of the fields is error prone. This could be resolved by storing the old pointer & capacity, running memcpy on the whole `CallbackBlock` and then restoring the pointer & capacity back.

Reply:
> However I agree that manual copying of the fields is error prone. This could be resolved by storing the old pointer & capacity, running memcpy on the whole CallbackBlock and then restoring the pointer & capacity back.

Ye. I wrongly thought, that the new capacity is zero.

Reply:
Resolved in 4393be0dc55e356bd2498c4e343273a1adfd41a6 :+1: